components/merkledb/benches/schema_patterns.rs

```diff
@@ -0,0 +1,327 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use criterion::{black_box, Bencher, Benchmark, Criterion, Throughput};
+use rand::{rngs::StdRng, Rng, SeedableRng};
+use serde_derive::{Deserialize, Serialize};
+
+use std::borrow::Cow;
+
+use exonum_crypto::Hash;
+use exonum_merkledb::{
+    access::{Access, AccessExt, Prefixed, RawAccessMut, Restore},
+    impl_object_hash_for_binary_value, BinaryValue, Database, Group, KeySetIndex, Lazy, MapIndex,
+    ObjectHash, ProofListIndex, ProofMapIndex, TemporaryDB,
+};
+
+const SEED: [u8; 32] = [100; 32];
+const SAMPLE_SIZE: usize = 10;
+const TX_COUNT: usize = 10_000;
+
+// Parameters used in transaction processing. See `EagerSchema` definition for context.
+/// Divisors used to form buckets for hot index group.
+const DIVISORS: &[u64] = &[23, 31, 47];
+/// Divisors used to form buckets for cold index group.
+const COLD_DIVISOR: u64 = 13;
+/// Chance to access `other_cold_index`.
+const COLD_CHANCE: u64 = 29;
+
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
+struct Transaction {
+    value: u64,
+    _payload: [u8; 32],
+}
+
+impl BinaryValue for Transaction {
+    fn to_bytes(&self) -> Vec<u8> {
+        bincode::serialize(self).unwrap()
+    }
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Result<Self, failure::Error> {
+        bincode::deserialize(bytes.as_ref()).map_err(From::from)
+    }
+}
+
+impl_object_hash_for_binary_value! { Transaction }
+
+trait ExecuteTransaction {
+    fn execute<T: Access>(fork: T, transaction: &Transaction)
+    where
+        T::Base: RawAccessMut;
+}
+
+struct EagerSchema<T: Access> {
+    // Accessed once per transaction.
+    transactions: MapIndex<T::Base, Hash, Transaction>,
+    // Hot index / group are accessed `DIVISORS.len()` times per transaction.
+    hot_index: ProofMapIndex<T::Base, u64, Hash>,
+    hot_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    // Cold index / group are accessed once per ~10 transactions.
+    cold_index: ProofMapIndex<T::Base, u64, Hash>,
+    cold_group: Group<T, u64, ProofListIndex<T::Base, u64>>,
+    // Accessed once per ~`COLD_DIVISOR` transactions.
+    other_cold_index: KeySetIndex<T::Base, u64>,
+}
+
+impl<T: Access> EagerSchema<T> {
+    fn new(access: T) -> Self {
+        Self {
+            transactions: Restore::restore(access.clone(), "transactions".into()).unwrap(),
+            hot_index: Restore::restore(access.clone(), "hot_index".into()).unwrap(),
```

Comment:
`access.clone()` is used really wide. Maybe make it `Copy`?

Reply:
`Prefixed` implements `Access` and is not `Copy`... See, e.g., its usage in `SnapshotExt` or `BlockchainData` - the prefix may be determined dynamically, so it is seemingly impossible to have `Prefixed` contain `&str` instead of `Cow<str>`. The only way I see it could work is to require a string service name (instead of more generic `InstanceQuery`) when accessing service data. Would this be OK?

Reply:
Hm. If `Prefixed` is the only reason that stops us here, maybe implement `Copy` for it manually?

Reply:
Not sure what you mean. Admittedly, I'm not very knowledgeable in the details of low-level memory management, but wouldn't implementing `Copy` for a type containing `String` lead to double memory frees?

Reply:
Also, I've been thinking about an approach in which `Prefixed` is replaced with *root objects*, as defined in the original MerkleDB spec. This would require substantially more changes in the MerkleDB interface though, which would hinder our release plans.

Reply:
>Not sure what you mean

Ah, indeed. I forgot that `Copy` trait cannot be implemented using some workaround. Sad.

>Also, I've been thinking about an approach in which Prefixed is replaced with root objects, as defined in the original MerkleDB spec. This would require substantially more changes in the MerkleDB interface though, which would hinder our release plans.

Yeah, I think it can wait.